From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Mon, 4 May 2026 15:37:21 +0800
Subject: [PATCH] Leaf: Optimize baby villager sensor

In the test, this can give ~16.58% improvement (~2316ms -> ~1932ms),
under 2048 villagers situation.
And ~42.93% improvement (~764ms -> ~436ms), under 512 villagers situation.

Co-authored by: Taiyou06 <kaandindar21@gmail.com>
As part of: Leaf (https://github.com/Winds-Studio/Leaf/blob/ver/1.21.11/leaf-server/minecraft-patches/features/0190-Optimize-baby-villager-sensor.patch)
Licensed under: GPL-3.0 (https://github.com/Winds-Studio/Leaf/blob/ver/1.21.11/licenses/GPL-3.0.txt)

diff --git a/net/minecraft/world/entity/ai/memory/NearestVisibleLivingEntities.java b/net/minecraft/world/entity/ai/memory/NearestVisibleLivingEntities.java
index 5c967b55f6ad3b660e9cdf74fadd90c5ea67afb9..7e30546c012533b7bd4ebb7ccaed30b978f0fe09 100644
--- a/net/minecraft/world/entity/ai/memory/NearestVisibleLivingEntities.java
+++ b/net/minecraft/world/entity/ai/memory/NearestVisibleLivingEntities.java
@@ -13,8 +13,8 @@ import net.minecraft.world.entity.ai.sensing.Sensor;
 
 public class NearestVisibleLivingEntities {
     private static final NearestVisibleLivingEntities EMPTY = new NearestVisibleLivingEntities();
-    private final List<LivingEntity> nearbyEntities;
-    private final Predicate<LivingEntity> lineOfSightTest;
+    public final List<LivingEntity> nearbyEntities; // Sewlia - to public // Leaf - to public
+    public final Predicate<LivingEntity> lineOfSightTest; // Sewlia - to public // Leaf - to public
 
     private NearestVisibleLivingEntities() {
         this.nearbyEntities = List.of();
diff --git a/net/minecraft/world/entity/ai/sensing/VillagerBabiesSensor.java b/net/minecraft/world/entity/ai/sensing/VillagerBabiesSensor.java
index e3b2089e74d57ae8a15e9e5d29f43054ec6ec295..fa23d161509058aab70a22c9635133df27681126 100644
--- a/net/minecraft/world/entity/ai/sensing/VillagerBabiesSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/VillagerBabiesSensor.java
@@ -22,7 +22,21 @@ public class VillagerBabiesSensor extends Sensor<LivingEntity> {
     }
 
     private List<LivingEntity> getNearestVillagerBabies(final LivingEntity myBody) {
-        return ImmutableList.copyOf(this.getVisibleEntities(myBody).findAll(this::isVillagerBaby));
+        // Leaf start - Optimize baby villager sensor
+        NearestVisibleLivingEntities visibleEntities = this.getVisibleEntities(myBody);
+        ImmutableList.Builder<LivingEntity> babies = ImmutableList.builder();
+
+        // Inline and use single loop - copy from NearestVisibleLivingEntities#findAll and isVillagerBaby
+        for (LivingEntity target : visibleEntities.nearbyEntities) {
+            if (target.getType() == EntityType.VILLAGER
+                    && target.isBaby()
+                    && visibleEntities.lineOfSightTest.test(target)) {
+                babies.add(target);
+            }
+        }
+
+        return babies.build();
+        // Leaf end - Optimize baby villager sensor
     }
 
     private boolean isVillagerBaby(final LivingEntity entity) {
